week7/answers.sql

```diff
@@ -1,3 +1,316 @@
+        
+CREATE OR REPLACE VIEW character_inv AS 
+	SELECT 
+			c.character_id AS character_id,
+			c.name AS character_name,
+			i.name AS item_name,
+			i.armor AS armor,
+			i.damage AS damage
+        FROM characters c
+        INNER JOIN inventory inv
+			ON inv.character_id = c.character_id
+        INNER JOIN items i
+			ON inv.item_id = i.item_id
+        GROUP BY c.character_id, inv.item_id
+        ORDER BY c.character_id ASC;
+        
+CREATE OR REPLACE VIEW character_items AS
+ 	SELECT 
+			ce.character_id AS character_id,
+			ce.character_name AS character_name,
+			ce.item_name AS item_name,
+			ce.armor AS armor,
+			ce.damage AS damage
+        FROM character_equipped ce
+		UNION
+	SELECT
+			ci.character_id AS character_id,
+			ci.character_name AS character_name,
+			ci.item_name AS item_name,
+			ci.armor AS armor,
+			ci.damage AS damage
+        FROM character_inv ci
+	ORDER BY item_name;
+	
+
+-- ----------------- TEAM INVENTORY CHECK ------------------
+CREATE OR REPLACE VIEW team_items AS 
+	SELECT 
+		t.team_id AS team_id,
+        t.name AS team_name,
+        i.name AS item_name,
+        i.armor AS armor,
+        i.damage AS damage
+        FROM teams t
+        INNER JOIN team_members tm
+		ON t.team_id = tm.team_id
+        INNER JOIN characters c
+		ON tm.character_id = c.character_id
+        INNER JOIN inventory inv
+		ON inv.character_id = c.character_id
+        INNER JOIN items i
+		ON i.item_id = inv.item_id
+GROUP BY t.team_id, inv.item_id;
+
+
+-- ------------------------- ARMOR TOTAL FUNCTION ----------------------------
+DELIMITER ;;
+CREATE FUNCTION armor_total(character_id INT UNSIGNED)
+    RETURNS INT UNSIGNED
+    DETERMINISTIC
+    BEGIN
+		
+        DECLARE total_armor INT UNSIGNED;
+        DECLARE stat_armor INT UNSIGNED;
+        DECLARE item_armor INT UNSIGNED;
+        
+		SELECT SUM(i.armor)
+			INTO item_armor 
+				FROM characters c
+				INNER JOIN equipped eq
+					ON c.character_id = eq.character_id
+				INNER JOIN items i
+					ON i.item_id = eq.item_id
+			WHERE c.character_id=character_id
+			GROUP BY c.character_id;
+		
+		SELECT cs.armor 
+			INTO stat_armor 
+				FROM characters c 
+				INNER JOIN character_stats cs 
+					ON c.character_id = cs.character_id 
+			WHERE c.character_id=character_id 
+			GROUP BY c.character_id;
+        
+        
+		IF item_armor > 0 THEN
+			SET total_armor = (item_armor + stat_armor);
+			RETURN total_armor;
+		ELSE 
+			RETURN stat_armor;
+		END IF;
+END;;
+DELIMITER ;
+    
+    
+-- ------------------------------- Attack Procedure -------------------------
+DELIMITER ;;
+CREATE PROCEDURE attack(target INT UNSIGNED, chosen_weapon INT UNSIGNED)
+BEGIN
+	
+    DECLARE outcome VARCHAR(32);
+    DECLARE char_armor TINYINT UNSIGNED;
+    DECLARE wep_damage TINYINT UNSIGNED;
+	DECLARE weapon TINYINT UNSIGNED;
+    DECLARE netdmg TINYINT SIGNED;
+    DECLARE char_health TINYINT SIGNED;
+    
+    SELECT armor_total(target) INTO char_armor;
```

Comment:
These three SELECT statements could be merged into a single SELECT with JOINs (noting, only).